File: leetcode/LC1/array/0121_best_time_to_buy_sell_stock.cpp
/* traverse from left to right, keeping track of minimum element and maximum profit
*/
//O(n)/O(1)

#include <climits>
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minimum = INT_MAX;
        int profit = 0;
        for(int i = 0; i < prices.size(); i++){
            if((prices[i] - minimum) > profit){
                profit = prices[i] - minimum;
            }
            if(prices[i] < minimum){
                minimum = prices[i];
            }
        }
        return profit;
    }
};

File: leetcode/LC/array/0121_best_time_to_buy_sell_stock.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0
        minimum = float('inf')
        
        for price in prices:
            if minimum > price:
                minimum = price
            currentProfit = price - minimum
            profit = max(currentProfit, profit)
        return profit
        